packages/neovim-common/src/ide/neovim/NeovimTextDocumentImpl.ts

```diff
@@ -0,0 +1,389 @@
+import {
+  EndOfLine,
+  Position,
+  Range,
+  TextDocument,
+  TextLine,
+} from "@cursorless/common";
+import { URI } from "vscode-uri";
+import NeovimTextLineImpl from "./NeovimTextLineImpl";
+
+export class NeovimTextDocumentImpl implements TextDocument {
+  private _uri: URI;
+  private _languageId: string;
+  private _version: number;
+  private _lineCount: number;
+  private _eol: string;
+
+  private _lines: string[];
+  private _lineStarts: PrefixSumComputer | null;
+  private _cachedTextValue: string | null;
+
+  get uri(): URI {
+    return this._uri;
+  }
+
+  get languageId(): string {
+    return this._languageId;
+  }
+
+  get version(): number {
+    return this._version;
+  }
+
+  get lineCount(): number {
+    // console.debug(`lineCount(): ${this._lineCount}`);
+    return this._lineCount;
+  }
+
+  get range(): Range {
+    const { end } = this.lineAt(this.lineCount - 1).range;
+    const range = new Range(0, 0, end.line, end.character);
+    // console.debug(
+    //   `range(): (${range.start.line},${range.start.character}),(${range.end.line},${range.end.character})`,
+    // );
+    return range;
+  }
+
+  get eol(): EndOfLine {
+    return this._eol === "\n" ? "LF" : "CRLF";
+  }
+
+  constructor(
+    uri: URI,
+    languageId: string,
+    version: number,
+    eol: string,
+    lines: string[],
+  ) {
+    this._uri = uri;
+    this._languageId = languageId;
+    this._version = version;
+    this._eol = eol;
+    this._lines = lines;
+
+    this._lineCount = lines.length;
+    this._lineStarts = null;
+    this._cachedTextValue = null;
+  }
+
+  public update(lines: string[]) {
+    this._lines = lines;
+
+    this._lineCount = lines.length;
+    this._lineStarts = null;
+    this._cachedTextValue = null;
+  }
+
+  public lineAt(lineOrPosition: number | Position): TextLine {
+    let line: number | undefined;
+    if (lineOrPosition instanceof Position) {
+      line = lineOrPosition.line;
+    } else if (typeof lineOrPosition === "number") {
+      line = lineOrPosition;
+    }
+    // console.debug(`lineAt() line=${line}`);
+
+    if (
+      typeof line !== "number" ||
+      line < 0 ||
+      line >= this._lines.length ||
+      Math.floor(line) !== line
+    ) {
+      throw new Error("Illegal value for `line`");
```

Comment:
It will just be clamped to the range. 

https://github.com/cursorless-dev/cursorless/blob/ace857e88a9eab37a2e7cfd3aa87082394d7722f/packages/cursorless-talon-js/src/ide/TalonJsTextDocument.ts#L48-L54